### PR TITLE
WT-10461 Fix key out of order in skip list on weakly ordered architecture

### PR DESCRIPTION
**Pull request by quchenhao**

### CODE REVIEW

**Comment by quchenhao** [comment]
@agorrod Here's the version just using ordered read.

**Comment by quchenhao** [comment]
WT performance test:
https://spruce.mongodb.com/version/63d994b81e2d170334851101/tasks?sorts=STATUS%3AASC%3BBASE_STATUS%3ADESC
Mongodb performance test:
https://spruce.mongodb.com/version/63d996203e8e86193aeb9849/tasks?sorts=STATUS%3AASC%3BBASE_STATUS%3ADESC

**Review by agorrod — APPROVED** [review]
Thanks Chenhao - this looks good to me. It's definitely more understandable than the other version.

The comments look good to me now, but could use scrutiny from others as well, since I now don't have a fresh perspective on their content.

File: src/btree/row_srch.c [hunk 1]
```diff
@@ -102,7 +102,18 @@ __wt_search_insert(
     match = skiphigh = skiplow = 0;
     ins = last_ins = NULL;
     for (i = WT_SKIP_MAXDEPTH - 1, insp = &ins_head->head[i]; i >= 0;) {
-        if ((ins = *insp) == NULL) {
+        /*
+         * The algorithm requires that the skip list insert pointer is not re-read during the loop.
+         * While the compiler can change the code in a way that it reads the ins value from memory
```
Comment (ruby-oujo) [inline src/btree/row_srch.c:107]: Starting this sentence with "While" is confusing. It would read better with "Whilst" or "Whereas", unless the previous sentence is joined together with this one.

Comment (ruby-oujo) [inline src/btree/row_srch.c:107]: Starting this sentence with "While" is confusing here. It would read better with "Whilst" or "Whereas", unless the previous sentence is joined together with this one.

File: src/btree/row_srch.c [hunk 2]
```diff
@@ -102,7 +102,18 @@ __wt_search_insert(
...
+         * again in the following code thus breaking this assumption.
+         *
+         * In addition, the CPU with weak memory ordering, such as ARM and PPC, may speculatively
```
Comment (ruby-oujo) [inline src/btree/row_srch.c:110]: I would replace "the CPU" with "a CPU".

File: src/btree/row_srch.c [hunk 3]
```diff
@@ -116,6 +127,40 @@ __wt_search_insert(
             last_ins = ins;
             key.data = WT_INSERT_KEY(ins);
             key.size = WT_INSERT_KEY_SIZE(ins);
+            /*
+             * We have an optimization to reduce the number of bytes we need to compare down the
+             * search if we know certain length of the prefix of the search key matches the keys we
+             * have already compared upper the stack. This works because we know the keys become
```
Comment (ruby-oujo) [inline src/btree/row_srch.c:133]: I suggest a little rephrasing here:  "We have an optimization to reduce the number of bytes we need to compare **during a** search if we know **a prefix** of the search key matches the keys we have already compared **on the upper stacks.** "

File: src/btree/row_srch.c [hunk 4]
```diff
@@ -116,6 +127,40 @@ __wt_search_insert(
...
+             *
+             * L1: AA -> AC -> BD
+             *
+             * L0: AA -> BD
+             *
+             * Since we have compared with AA and AC on level 1 before dropping down to level 0, we
+             * decide we can skip comparing the first byte of the key. However, since we don't see
+             * AC on level 0, we compare with BD and wrongly skip the comparison with prefix B.
+             *
+             * On the architectures with strong memory ordering, the requirement is satisfied by
```
Comment (ruby-oujo) [inline src/btree/row_srch.c:158]: remove "the" from "the architectures".

File: src/include/serial_inline.h [hunk 5]
```diff
@@ -29,7 +30,15 @@ __insert_simple_func(
      * implementations read the old value multiple times.
      */
     for (i = 0; i < skipdepth; i++) {
-        WT_INSERT *old_ins = *ins_stack[i];
+        /*
+         * The insert stack position must be read only once - if the compiler chose to re-read the
```
Comment (ruby-oujo) [inline src/include/serial_inline.h:34]: replace "chose" with "chooses"

File: src/include/serial_inline.h [hunk 6]
```diff
@@ -63,7 +73,15 @@ __insert_serial_func(WT_SESSION_IMPL *session, WT_INSERT_HEAD *ins_head, WT_INSE
...
```
Comment (ruby-oujo) [inline src/include/serial_inline.h:77]: again here, "chose" to "chooses".

**Review by ruby-oujo — APPROVED** [review]
LGTM, just left a few comments.

**Review by korteland — APPROVED** [review]
LGTM! The comments are great, thanks for the effort there.

File: src/btree/row_srch.c [hunk 7]
```diff
@@ -116,6 +127,40 @@ __wt_search_insert(
...
+             *
+             * L1: AA -> AC -> BD
+             *
+             * L0: AA -> BD
+             *
+             * Since we have compared with AA and AC on level 1 before dropping down to level 0, we
+             * decide we can skip comparing the first byte of the key. However, since we don't see
+             * AC on level 0, we compare with BD and wrongly skip the comparison with prefix B.
+             *
+             * On architectures with strong memory ordering, the requirement is satisfied by
```
Comment (sulabhM) [inline src/btree/row_srch.c:158]: This might need a little bit more elaboration or a correction. ARM's ```CASAL``` is also a full barrier comparable to x86_64's ```LOCK CMPXCHG```. What is different is that reads are not re-ordered against other reads on x86, whereas on weaker memory models they could be. In both cases, a full barrier would clear the cache line and synchronize it to the latest values. But that doesn't prevent us on ARM from having already read L0 before the CAS operation and then having the "stale" information.

File: src/btree/row_srch.c [hunk 8]
```diff
@@ -102,7 +102,18 @@ __wt_search_insert(
...
-        if ((ins = *insp) == NULL) {
+        /*
+         * The algorithm requires that the skip list insert pointer is not re-read during the loop.
+         * Whilst the compiler can change the code in a way that it reads the ins value from memory
+         * again in the following code thus breaking this assumption.
+         *
+         * In addition, a CPU with weak memory ordering, such as ARM and PPC, may speculatively read
+         * an old value. It is not OK and the reason is explained in the following comment.
+         *
+         * Place a read barrier here using atomic read to avoid these issues.
```
Comment (sulabhM) [inline src/btree/row_srch.c:113]: ```WT_ORDERED_READ``` uses fence instructions to issue a read barrier; it is not an atomic read. This line needs to be fixed.

File: src/btree/row_srch.c [hunk 9]
```diff
@@ -102,7 +102,18 @@ __wt_search_insert(
     match = skiphigh = skiplow = 0;
     ins = last_ins = NULL;
     for (i = WT_SKIP_MAXDEPTH - 1, insp = &ins_head->head[i]; i >= 0;) {
-        if ((ins = *insp) == NULL) {
+        /*
```
Comment (sulabhM) [inline src/btree/row_srch.c:105]: Is it clearer if we say that our intention here is to cache the copy of the insert pointer at the start of the loop iteration? This ensures we do not re-read a conflicting value if it were to change during the iteration. The read barrier also prevents the compiler from optimising the cached value away.

For instance, we could say:

"Do an ordered read to get the insert pointer. A CPU with weak memory ordering, such as ARM and PPC, may re-order insert pointer reads. This might lead to incorrectly using an older value. It is not OK, and the reason is explained in the following comment. We also cache the insert pointer at the beginning of the loop iteration. This ensures we don't re-read a conflicting value if it were to change during the iteration. Using a read barrier also prevents the compiler from optimising away the cached value."

File: src/include/serial_inline.h [hunk 10]
```diff
@@ -29,7 +30,15 @@ __insert_simple_func(
      * implementations read the old value multiple times.
      */
     for (i = 0; i < skipdepth; i++) {
-        WT_INSERT *old_ins = *ins_stack[i];
+        /*
+         * The insert stack position must be read only once - if the compiler chooses to re-read the
```
Comment (sulabhM) [inline src/include/serial_inline.h:34]: If you choose my suggested re-wording of the comment above, you can also use similar words here for consistency.

File: src/include/serial_inline.h [hunk 11]
```diff
@@ -63,7 +73,15 @@ __insert_serial_func(WT_SESSION_IMPL *session, WT_INSERT_HEAD *ins_head, WT_INSE
      * implementations read the old value multiple times.
      */
     for (i = 0; i < skipdepth; i++) {
-        WT_INSERT *old_ins = *ins_stack[i];
+        /*
```
Comment (sulabhM) [inline src/include/serial_inline.h:76]: If you choose my suggested re-wording of the comment above, you can also use similar words here for consistency.

**Review by sulabhM — APPROVED** [review]
I have suggested a re-wording of the comments for my clarity. But I will leave it up to you to accept or reject these suggestions.

The change itself LGTM!

File: src/btree/row_srch.c [hunk 12]
```diff
@@ -116,6 +127,42 @@ __wt_search_insert(
...
+             * the stack.
+             *
+             * However, things become tricky if we have another key inserted concurrently to next of
+             * the search key. If we have seen this key inserted upper the stack, we require the
+             * insertion must also be visible to us down the stack. Otherwise, we may wrongly skip
+             * the comparison of a prefix and land on the wrong spot. Here's an example:
+             *
+             * Suppose we have a skip list:
+             *
+             * L1: AA -> BD
```
Comment (dgomezferro) [inline src/btree/row_srch.c:143]: Could you use `BA` as the second key? That way it fully exemplifies the bug, since then we compare A**B** to B**A** and insert it out of order.

File: src/btree/row_srch.c [hunk 13]
```diff
@@ -116,6 +127,42 @@ __wt_search_insert(
...
             key.data = WT_INSERT_KEY(ins);
             key.size = WT_INSERT_KEY_SIZE(ins);
+            /*
+             * We have an optimization to reduce the number of bytes we need to compare during the
+             * search if we know a prefix of the search key matches the keys we have already
+             * compared on the upper stacks. This works because we know the keys become denser down
+             * the stack.
+             *
+             * However, things become tricky if we have another key inserted concurrently to next of
+             * the search key. If we have seen this key inserted upper the stack, we require the
```
Comment (dgomezferro) [inline src/btree/row_srch.c:137]:
```suggestion
             * However, things become tricky if we have another key inserted concurrently right next
             * to the search key. If we have seen this key inserted upper the stack, we require the
```

**Review by dgomezferro — COMMENTED** [review]
LGTM, just two minor comments

**Review by dgottlieb — APPROVED** [review]

File: src/btree/row_srch.c (hunk 7 above)
Comment (RedBeard0531) [reply src/btree/row_srch.c:158]: > a full barrier would clear the cache line and synchronize it to the latest values

That isn't true at all. A) all modern CPUs are fully cache coherent using something like [MOESI](https://en.wikipedia.org/wiki/MOESI_protocol), so if a cache read succeeds, you are guaranteed to get its "latest value" (I'm ignoring things like mapped mapped I/O and non-CPU devices writing to memory, since that is more a concern for kernels than us). Barriers don't have to do any sort of cache-line clearing, since that is done automatically as part of normal reads and writes. The issue is that superscalar cores (which is basically of them at this point) may have multiple reads or writes in flight at once and without fencing, they may complete in an order that doesn't match the program order. 

Note that even x86 isn't fully ordered and allows loads to complete prior to (plain) stores that are earlier in program order (Run [this](https://gist.github.com/RedBeard0531/b8b0a7b41e7bdaddbbddb9d226df2452#file-sc-cpp-L88-L101) on x86 if you want proof). This is due to store buffers which allow a write instruction to "complete" before it makes it to cache and becomes globally visible.

But none of this really matters. We shouldn't be talking about the hardware implementation, since that isn't part of the platform guarantees. We should instead be able to justify correctness by pointing to what we do to ensure that the memory model guarantees us correct ordering  where we need it. In this case, we need to read the L(n+1) pointer prior to reading the L(n) pointer. In the old code, we had a bug because we did not ensure that the reads were ordered correctly and the fix is to correctly order the reads. I think the only thing that needs explaining is _why_ we care about the order. We should accept that `WT_ORDERED_READ` correctly enforces ordering and take to heart that we must always do _something_ to enforce order for both reads and writes whenever we care about it. We shouldn't need to talk about strong and weak memory models and specific instructions in general code because it shouldn't care about that. It just needs to use our preferred mechanisms to ensure that ordering.

File: src/btree/row_srch.c [hunk 14]
```diff
@@ -102,7 +102,18 @@ __wt_search_insert(
...
+         * The algorithm requires that the skip list insert pointer is only read once within the
+         * loop. Whilst the compiler can change the code in a way that it reads the insert pointer
+         * value from memory again in the following code.
+         *
+         * In addition, a CPU with weak memory ordering, such as ARM, may speculatively read an old
+         * value. It is not OK and the reason is explained in the following comment.
+         *
+         * Place a read barrier here to avoid these issues.
+         */
+        WT_ORDERED_READ(ins, *insp);
```
Comment (RedBeard0531) [inline src/btree/row_srch.c:115]: Does `*insp` need to be volatile? Otherwise what prevents torn reads (and other fun things)?

We should probably put something like [this](https://elixir.bootlin.com/linux/latest/source/include/asm-generic/rwonce.h#L44) from linux's `READ_ONCE` into `WT_ORDERED_READ` to add `volatile` qualification when doing the write to tell the compiler that it is important that this is done precisely as we say here.

(I am fully aware that `volatile` is not spec-correct for multi-threading, but since we aren't moving to the C11 model in this commit, we should at least do what we can to tell the compiler that this read is "special")

Comment (keitharnoldsmith) [reply src/btree/row_srch.c:115]: > Does *insp need to be volatile? Otherwise what prevents torn reads (and other fun things)?

WiredTiger [requires](https://github.com/wiredtiger/wiredtiger/blob/32ee82a40f6d58579ed28ce5eb979a9fbda391df/src/include/gcc.h#L53) atomic reads and writes of 64-bit memory locations. So we don't take additional steps to ensure atomicity (i.e., prevent torn reads/writes) when loading and storing pointer values to shared memory.

Is there reason to believe that the platforms we run on don't provide this guarantee?

File: src/btree/row_srch.c [hunk 15]
```diff
@@ -102,7 +102,18 @@ __wt_search_insert(
     match = skiphigh = skiplow = 0;
     ins = last_ins = NULL;
     for (i = WT_SKIP_MAXDEPTH - 1, insp = &ins_head->head[i]; i >= 0;) {
-        if ((ins = *insp) == NULL) {
+        /*
+         * The algorithm requires that the skip list insert pointer is only read once within the
+         * loop. Whilst the compiler can change the code in a way that it reads the insert pointer
```
Comment (keitharnoldsmith) [inline src/btree/row_srch.c:107]: Extremely minor nit: I would just say "While" instead of using the more archaic "Whilst".

File: src/btree/row_srch.c (hunk 14 above)
Comment (amschwerin) [reply src/btree/row_srch.c:115]: Are you sure that you're getting a read every time, and that it's never optimized away?

File: src/btree/row_srch.c [hunk 16]
```diff
@@ -116,6 +127,42 @@ __wt_search_insert(
             last_ins = ins;
             key.data = WT_INSERT_KEY(ins);
             key.size = WT_INSERT_KEY_SIZE(ins);
+            /*
+             * We have an optimization to reduce the number of bytes we need to compare during the
+             * search if we know a prefix of the search key matches the keys we have already
+             * compared on the upper stacks. This works because we know the keys become denser down
+             * the stack.
+             *
+             * However, things become tricky if we have another key inserted concurrently to next of
```
Comment (keitharnoldsmith) [inline src/btree/row_srch.c:136]: > However, things become tricky if we have another key ... we may wrongly skip the comparison ... and land on the wrong spot.

I think this framing is backwards. It suggests that we update the pointers in the stack in order to avoid breaking  the prefix optimization.

I would describe this by saying that the insert is designed so any concurrent reader will alway see a valid skip list. I.e., at any level, of the list, keys are in sorted order, and that if a reader sees a key in level N, that key is also in levels below N. 

In this framing, the WT-10461 bug is an example of how code (the prefix optimization) that depends on seeing a valid skip list can break if that guarantee is violated.

The value of describing it this way is that it makes it clear the skip list invariants are required independent of the prefix optimization.

File: src/btree/row_srch.c (hunk 14 above)
Comment (keitharnoldsmith) [reply src/btree/row_srch.c:115]: > Are you sure that you're getting a read every time, and that it's never optimized away?

I was responding specifically to the concern about torn reads. So fair point -- using a `volatile` would also prevent "other fun things" by ensuring that we really load a value from memory.

In this case, however, I can't see it. The barrier in the loop ensures that each iteration of the loop must load a new value of `ins` from memory. This code is not updating that information in memory, so the compiler has no way to know what is in that memory without reading it. If the compiler has an oracle that can accurately determine what is in memory at some point between two barriers, then it is giving us a correct value! :-)

I think @RedBeard0531's question about including a `volatile` qualifier in the `WT_ORDERED_READ()` macro is a reasonable question/proposal for what we might want to do in WT in the future. So in this thread I'm basically arguing that it is sufficient and desirable to just make the smallest change that will fix the issue in WT-10461 so we minimize the chances of inadvertently breaking other stuff. Happy to follow up about broader improvements in subsequent tickets.

**Comment by amschwerin** [comment]
Ack.

File: src/btree/row_srch.c (hunk 14 above)
Comment (hanumantmk) [reply src/btree/row_srch.c:115]: For what it's worth, I went and dug up a 2019 Linux kernel thread specifically talking about torn reads on aligned memory locations.  Where they had seen it, which partially drives the use of volatile in the kernel.

https://lore.kernel.org/all/CAHk-=wj2t+GK+DGQ7Xy6U7zMf72e7Jkxn4_-kGyfH3WFEoH+YQ@mail.gmail.com/T/#u

> We do require a _minimum_ of compiler sanity, but the compiler turning a non-marked read into two reads has definitely happened, and can be very subtle. Even if on a C level it looks like a single access, and correctness doesn't care about _what_ the value we read is, it might be turned by the compiler into two separate accesses that get two different values, and then the end result may be insane and unreliable.

And a link to all the UB we're inviting in with bare loads and stores: https://lwn.net/Articles/793253/

I don't know if it's in this commit, but at some point we've got to move away from relying so much on undefined behavior.

File: src/btree/row_srch.c [hunk 17]
```diff
@@ -102,7 +102,18 @@ __wt_search_insert(
...
     ins = last_ins = NULL;
     for (i = WT_SKIP_MAXDEPTH - 1, insp = &ins_head->head[i]; i >= 0;) {
-        if ((ins = *insp) == NULL) {
+        /*
+         * The algorithm requires that the skip list insert pointer is only read once within the
+         * loop. Whilst the compiler can change the code in a way that it reads the insert pointer
+         * value from memory again in the following code.
+         *
+         * In addition, a CPU with weak memory ordering, such as ARM, may speculatively read an old
+         * value. It is not OK and the reason is explained in the following comment.
```
Comment (RedBeard0531) [inline src/btree/row_srch.c:111]: The situation is actually worse than this comment is saying. While it may sound weird, without a barrier, there is nothing ensuring that when we dereference the skiplist pointer that the data we get back is newer than when the pointer was written. This includes the key and anything we return to the caller. In assembly this is less of an issue because arm (like all processors other than alpha) honor something called an "address dependency" to ensure that if you read an address that was provably written after the data it  points to (using something like WT_PUBLISH, or a CAS with at lease release semantics) then all reads from the address will happen after the load of the address itself. You can see this described briefly at the bottom of [this page of the arm docs](https://developer.arm.com/documentation/102336/0100/Memory-ordering).

And that would be great if we were writing assembly. The intent was that the C memory model would provide access to this (nearly) universal hardware guarantee using the "consume" memory ordering. Unfortunately compiler venders have found it nearly impossible to ensure that address dependencies from high level source code are always carried through correctly to the generated assembly, and all of them now treat it identically to "acquire", which thankfully is really cheap. While sometimes you may get lucky (and we probably have so far), it is really best not to rely on this unless we want to start auditing assembly before every release.

TL;DR any time you are loading a pointer stored in a concurrently updated data structure, which this code is doing, you almost certainly need acquire semantics or `WT_ORDERED_READ`. This would apply even if this code was for a linked list with a single pointer per node, rather than a skiplist with multiple pointers.

File: src/btree/row_srch.c (hunk 14 above)
Comment (keitharnoldsmith) [reply src/btree/row_srch.c:115]: Thanks @hanumantmk. That's fascinating! (Also a bit frustrating as I'd really like to know when/where/how they found read tearing here. But Linus is silent on that...)

File: src/btree/row_srch.c (hunk 15 above)
Comment (ruby-oujo) [reply src/btree/row_srch.c:107]: @keitharnoldsmith , I actually requested this to read either "Whilst" or "Whereas," or just to join the two sentences, but this was my mistake. As these are subordinate conjunctions, it can be misleading to have only one clause in the sentence, especially if we start with the sentence with them. I think the better solution here would have been to just join the two sentences. That was my mistake, sorry @quchenhao!

File: src/btree/row_srch.c (hunk 12 above)
Comment (quchenhao) [reply src/btree/row_srch.c:143]: Good point. That's a mistake. Fixed

File: src/btree/row_srch.c (hunk 16 above)
Comment (quchenhao) [reply src/btree/row_srch.c:136]: This is a good idea.

File: src/btree/row_srch.c (hunk 7 above)
Comment (quchenhao) [reply src/btree/row_srch.c:158]: I removed the word about detailed hardware implementation. But I think we still need to mention about the hardware architecture because wiretiger is not supporting all the hardware platforms and we need to be aware of there differences.

File: src/btree/row_srch.c (hunk 17 above)
Comment (quchenhao) [reply src/btree/row_srch.c:111]: got it.

File: src/btree/row_srch.c (hunk 7 above)
Comment (sulabhM) [reply src/btree/row_srch.c:164]: @RedBeard0531 , thanks for giving more insight into cache coherence on modern CPUs. I agree that talking about hardware implementations is a distraction we should stay away from. Until we follow the memory access order required for correctness, we don't need to be any wiser about how each platform manages that.

**Review by jeremythorp — APPROVED** [review]
Thanks @quchenhao, that LGTM!

**Review by keitharnoldsmith — APPROVED** [review]
LGTM

**Review by pmacko86 — COMMENTED** [review]
Just wondering, is there a possibility that we may need a read barrier in __wt_search_insert in the inner for loop that's taken when cmp == 0 (around line 182 of row_srch.c)? I don't think this specific reordering happens there, but I'm just trying to wrap my head around the possibility of some reordering across the loop iterations and whether that could cause any problem. Just thinking aloud :)

But overall, these changes LGTM!

**Comment by dgomezferro** [comment]
@pmacko86 I think you are right, without a read barrier we can copy inconsistent state into `cbt->next_stack` (a newer pointer in the upper layer, an older pointer in the lower layer). However, I don't think we use `cbt->next_stack` later in that case, when we have a match.

**Comment by quchenhao** [comment]
@pmacko86 @dgomezferro I agree it is a problem and I think we need to fix that even though next_stack is not used later. Also I realized that the column store search code have the same issue and I will fix as well.

**Comment by dgomezferro** [comment]
@quchenhao if it's not used would it be better not to populate it, or would that break the contract of the function / data structure?

**Comment by quchenhao** [comment]
> @quchenhao if it's not used would it be better not to populate it, or would that break the contract of the function / data structure?

Yes, I think it breaks the contract that we should have a valid and consistent position for next_stack. I'm worried that if we leave this unfixed, in the future we may start to use next stack when we find a match and the problem will resurface. Better to enforce that now.
